// discord-player-plus/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.de = exports.en = exports.createSetVolumeCommand = exports.createStopCommand = exports.createSongCommand = exports.createSkipCommand = exports.createShuffleCommand = exports.createSeekCommand = exports.createResumeCommand = exports.createRepeatCommand = exports.createQueueCommand = exports.createPlayCommand = exports.createPauseCommand = exports.createHelpCommand = exports.createClearCommand = exports.createAddCommand = exports.handleSlashCommand = void 0;
var commands_1 = require("./commands");
Object.defineProperty(exports, "handleSlashCommand", { enumerable: true, get: function () { return commands_1.handleSlashCommand; } });
var add_1 = require("./commands/add");
Object.defineProperty(exports, "createAddCommand", { enumerable: true, get: function () { return add_1.createAddCommand; } });
var clear_1 = require("./commands/clear");
Object.defineProperty(exports, "createClearCommand", { enumerable: true, get: function () { return clear_1.createClearCommand; } });
var help_1 = require("./commands/help");
Object.defineProperty(exports, "createHelpCommand", { enumerable: true, get: function () { return help_1.createHelpCommand; } });
var pause_1 = require("./commands/pause");
Object.defineProperty(exports, "createPauseCommand", { enumerable: true, get: function () { return pause_1.createPauseCommand; } });
var play_1 = require("./commands/play");
Object.defineProperty(exports, "createPlayCommand", { enumerable: true, get: function () { return play_1.createPlayCommand; } });
var queue_1 = require("./commands/queue");
Object.defineProperty(exports, "createQueueCommand", { enumerable: true, get: function () { return queue_1.createQueueCommand; } });
var repeat_1 = require("./commands/repeat");
Object.defineProperty(exports, "createRepeatCommand", { enumerable: true, get: function () { return repeat_1.createRepeatCommand; } });
var resume_1 = require("./commands/resume");
Object.defineProperty(exports, "createResumeCommand", { enumerable: true, get: function () { return resume_1.createResumeCommand; } });
var seek_1 = require("./commands/seek");
Object.defineProperty(exports, "createSeekCommand", { enumerable: true, get: function () { return seek_1.createSeekCommand; } });
var shuffle_1 = require("./commands/shuffle");
Object.defineProperty(exports, "createShuffleCommand", { enumerable: true, get: function () { return shuffle_1.createShuffleCommand; } });
var skip_1 = require("./commands/skip");
Object.defineProperty(exports, "createSkipCommand", { enumerable: true, get: function () { return skip_1.createSkipCommand; } });
var song_1 = require("./commands/song");
Object.defineProperty(exports, "createSongCommand", { enumerable: true, get: function () { return song_1.createSongCommand; } });
var stop_1 = require("./commands/stop");
Object.defineProperty(exports, "createStopCommand", { enumerable: true, get: function () { return stop_1.createStopCommand; } });
var volume_1 = require("./commands/volume");
Object.defineProperty(exports, "createSetVolumeCommand", { enumerable: true, get: function () { return volume_1.createSetVolumeCommand; } });
__exportStar(require("./player"), exports);
__exportStar(require("./player-manager"), exports);
__exportStar(require("./types/commands"), exports);
__exportStar(require("./types/engines"), exports);
__exportStar(require("./types/player"), exports);
__exportStar(require("./utils/player"), exports);
const de_json_1 = __importDefault(require("./languages/de.json"));
exports.de = de_json_1.default;
const en_json_1 = __importDefault(require("./languages/en.json"));
exports.en = en_json_1.default;
