// discord-player-plus/lib/commands/queue.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createQueueCommand = void 0;
const player_1 = require("../utils/player");
/**
 * Creates a `/queue` command for displaying queued tracks.
 */
const createQueueCommand = (playerManager, options) => {
    return {
        name: "queue",
        description: playerManager.translations.queue.description,
        run: (interaction) => __awaiter(void 0, void 0, void 0, function* () {
            var _a, _b;
            const player = playerManager.find(interaction.guildId);
            if (!player) {
                yield interaction.reply({
                    content: playerManager.translations.global.noGuildPlayer,
                    ephemeral: (_a = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _a !== void 0 ? _a : true,
                });
                return false;
            }
            const currentTrack = player.getCurrentTrack();
            const tracks = player.getQueue();
            if (!currentTrack || !tracks.length) {
                yield interaction.reply({
                    content: playerManager.translations.queue.empty,
                    ephemeral: (_b = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _b !== void 0 ? _b : true,
                });
                return false;
            }
            let content = `>>> **Queue (${tracks.length})**`;
            content += `\n\n**${playerManager.translations.queue.currentTrack}**: ${(0, player_1.trackToMarkdown)(currentTrack, true)}`;
            for (let i = 0; i < tracks.length; i++) {
                let newContent = i > 0 ? "\n" : "\n\n";
                newContent += `\`${i + 1}\`: ${(0, player_1.trackToMarkdown)(tracks[i], true)}`;
                // discord only allows max 2000 chars
                if (content.length + newContent.length <= 1995) {
                    content += newContent;
                    continue;
                }
                if (!content.endsWith("\n..."))
                    content += "\n...";
                break;
            }
            yield interaction.reply({ content, ephemeral: options === null || options === void 0 ? void 0 : options.ephemeral });
            return true;
        }),
    };
};
exports.createQueueCommand = createQueueCommand;
