// discord-player-plus/lib/commands/resume.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResumeCommand = void 0;
/**
 * Creates a `/resume` command for resuming the currently paused track.
 */
const createResumeCommand = (playerManager, options) => {
    return {
        name: "resume",
        description: playerManager.translations.resume.description,
        run: (interaction) => __awaiter(void 0, void 0, void 0, function* () {
            var _a, _b;
            const player = playerManager.find(interaction.guildId);
            if (!player) {
                yield interaction.reply({
                    content: playerManager.translations.global.noGuildPlayer,
                    ephemeral: (_a = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _a !== void 0 ? _a : true,
                });
                return false;
            }
            const paused = player.setPause(false);
            if (!paused) {
                yield interaction.reply({
                    content: playerManager.translations.resume.failure,
                    ephemeral: (_b = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _b !== void 0 ? _b : true,
                });
                return false;
            }
            yield interaction.reply({
                content: playerManager.translations.resume.success,
                ephemeral: options === null || options === void 0 ? void 0 : options.ephemeral,
            });
            return true;
        }),
    };
};
exports.createResumeCommand = createResumeCommand;
