// discord-player-plus/lib/commands/seek.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSeekCommand = void 0;
const discord_js_1 = require("discord.js");
const player_1 = require("../utils/player");
/**
 * Creates a `/seek` command for seeking the current track to a specific duration.
 */
const createSeekCommand = (playerManager, options) => {
    return {
        name: "seek",
        description: playerManager.translations.seek.description,
        options: [
            {
                name: "time",
                description: playerManager.translations.seek.optionDescription,
                type: discord_js_1.ApplicationCommandOptionType.Number,
                required: true,
                minValue: 0,
            },
        ],
        run: (interaction) => __awaiter(void 0, void 0, void 0, function* () {
            var _a;
            const player = playerManager.find(interaction.guildId);
            if (!player) {
                yield interaction.reply({
                    content: playerManager.translations.global.noGuildPlayer,
                    ephemeral: (_a = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _a !== void 0 ? _a : true,
                });
                return false;
            }
            yield interaction.deferReply({ ephemeral: options === null || options === void 0 ? void 0 : options.ephemeral });
            const time = interaction.options.getNumber("time", true);
            const success = yield player.seek(time * 1000);
            if (!success) {
                yield interaction.followUp({
                    content: playerManager.translations.seek.failure,
                });
                return false;
            }
            yield interaction.followUp({
                content: playerManager.translations.seek.success.replace("{duration}", (0, player_1.formatDuration)(time)),
                ephemeral: options === null || options === void 0 ? void 0 : options.ephemeral,
            });
            return true;
        }),
    };
};
exports.createSeekCommand = createSeekCommand;
