// discord-player-plus/lib/commands/skip.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSkipCommand = void 0;
const player_1 = require("../utils/player");
/**
 * Creates a `/skip` command for skipping the current track.
 */
const createSkipCommand = (playerManager, options) => {
    return {
        name: "skip",
        description: playerManager.translations.skip.description,
        run: (interaction) => __awaiter(void 0, void 0, void 0, function* () {
            var _a, _b;
            const player = playerManager.find(interaction.guildId);
            if (!player) {
                yield interaction.reply({
                    content: playerManager.translations.global.noGuildPlayer,
                    ephemeral: (_a = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _a !== void 0 ? _a : true,
                });
                return false;
            }
            const track = player.skip();
            if (!track) {
                yield interaction.reply({
                    content: playerManager.translations.skip.failure,
                    ephemeral: (_b = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _b !== void 0 ? _b : true,
                });
                return false;
            }
            yield interaction.reply({
                content: playerManager.translations.skip.success.replace("{song}", (0, player_1.trackToMarkdown)(track, true)),
                ephemeral: options === null || options === void 0 ? void 0 : options.ephemeral,
            });
            return true;
        }),
    };
};
exports.createSkipCommand = createSkipCommand;
