// discord-player-plus/lib/commands/volume.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSetVolumeCommand = void 0;
const discord_js_1 = require("discord.js");
/**
 * Creates a `/setvolume` command for setting the player volume between 0 and 200.
 */
const createSetVolumeCommand = (playerManager, options) => {
    return {
        name: "setvolume",
        description: playerManager.translations.setvolume.description,
        options: [
            {
                name: "volume",
                description: playerManager.translations.setvolume.optionDescription,
                type: discord_js_1.ApplicationCommandOptionType.Number,
                required: true,
                minValue: 0,
                maxValue: 200,
            },
        ],
        run: (interaction) => __awaiter(void 0, void 0, void 0, function* () {
            var _a, _b;
            const player = playerManager.find(interaction.guildId);
            if (!player) {
                yield interaction.reply({
                    content: playerManager.translations.global.noGuildPlayer,
                    ephemeral: (_a = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _a !== void 0 ? _a : true,
                });
                return false;
            }
            const volume = interaction.options.getNumber("volume", true);
            const success = player.setVolume(volume);
            if (!success) {
                yield interaction.reply({
                    content: playerManager.translations.setvolume.failure,
                    ephemeral: (_b = options === null || options === void 0 ? void 0 : options.ephemeralError) !== null && _b !== void 0 ? _b : true,
                });
                return false;
            }
            yield interaction.reply({
                content: playerManager.translations.setvolume.success.replace("{volume}", volume.toString()),
                ephemeral: options === null || options === void 0 ? void 0 : options.ephemeral,
            });
            return true;
        }),
    };
};
exports.createSetVolumeCommand = createSetVolumeCommand;
