// discord-player-plus/lib/types/player.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerError = exports.PlayerErrorCode = exports.PlayerRepeatMode = void 0;
// #endregion PlayerManagerOptions
// #region PlayerRepeatMode
var PlayerRepeatMode;
(function (PlayerRepeatMode) {
    /** No tracks are repeated (default) */
    PlayerRepeatMode[PlayerRepeatMode["NONE"] = 0] = "NONE";
    /** Repeat currently playing track */
    PlayerRepeatMode[PlayerRepeatMode["TRACK"] = 1] = "TRACK";
})(PlayerRepeatMode = exports.PlayerRepeatMode || (exports.PlayerRepeatMode = {}));
// #endregion PlayerRepeatMode
var PlayerErrorCode;
(function (PlayerErrorCode) {
    PlayerErrorCode[PlayerErrorCode["REFUSED_TO_SWITCH_VOICE_CHANNEL"] = 0] = "REFUSED_TO_SWITCH_VOICE_CHANNEL";
    PlayerErrorCode[PlayerErrorCode["UNKNOWN_PLAYER_ENGINE"] = 1] = "UNKNOWN_PLAYER_ENGINE";
    PlayerErrorCode[PlayerErrorCode["UNABLE_TO_CREATE_TRACK_STREAM"] = 2] = "UNABLE_TO_CREATE_TRACK_STREAM";
    PlayerErrorCode[PlayerErrorCode["INITIAL_VOLUME_FUNCTION_ERROR"] = 3] = "INITIAL_VOLUME_FUNCTION_ERROR";
})(PlayerErrorCode = exports.PlayerErrorCode || (exports.PlayerErrorCode = {}));
class PlayerError extends Error {
    constructor(code, msg) {
        super(msg);
        this.code = code;
        // Set the prototype explicitly.
        Object.setPrototypeOf(this, PlayerError.prototype);
    }
}
exports.PlayerError = PlayerError;
